src/query_graph/graph_path.rs

```diff
@@ -1405,12 +1407,101 @@ impl OpGraphPath {
+        itf: InterfaceFieldDefinitionPosition,
     ) -> Result<bool, FederationError> {
-        todo!()
+        let valid_schema = self.graph.schema()?;
+        let schema = valid_schema.schema();
+        let fed_spec = get_federation_spec_definition_from_subgraph(valid_schema)?;
+        let key_directive = fed_spec.key_directive_definition(valid_schema)?;
+        let shareable_directive = fed_spec.shareable_directive(valid_schema)?;
+        let comp_type_pos = CompositeTypeDefinitionPosition::Interface(itf.parent());
+        for implem in valid_schema.possible_runtime_types(comp_type_pos)? {
+            let field = schema
+                .get_interface(&implem.type_name)
+                .and_then(|n| n.fields.get(&itf.field_name))
+                .ok_or_else(|| {
+                    FederationError::internal(
+                        "Unable to find interface field ({itf}) in schema: {schema}",
+                    )
+                })?;
+            let ty = valid_schema
+                .get_type(implem.type_name.clone())?
+                .get(schema)?;
+            if !ty.directives().has(&key_directive.name) {
+                continue;
+            }
+            if !field.directives.has(&shareable_directive.name) {
+                continue;
+            }
+            let base_ty_name = field.ty.inner_named_type();
+            if is_leaf_type(schema, base_ty_name) {
+                continue;
+            }
+            let Some(ty) = schema.get_object(base_ty_name) else {
+                return Ok(true);
+            };
+            if ty
+                .fields
+                .values()
+                .any(|f| !is_leaf_type(schema, f.ty.inner_named_type()))
+            {
+                return Ok(true);
+            }
+            for node in self.graph.nodes_for_type(&ty.name) {
+                let node = self.graph.node_weight(node)?;
+                let tail = self.graph.node_weight(self.tail)?;
+                if node.source == tail.source {
+                    continue;
+                }
+                // PORT_NOTE: The JS code pulls out `otherMetadata` from each node here. This does
+                // not seem to be needed in the RS code?
+                let build_err = || {
+                    Err(FederationError::internal(format!(
+                        "{implem} is an object in {} but a {} in {}",
+                        tail.source, node.type_, node.source
+                    )))
+                };
+                let QueryGraphNodeType::SchemaType(node_ty) = &node.type_ else {
+                    return build_err();
+                };
+                let node_ty_name = node_ty.type_name();
+                let node_ty = valid_schema.get_type(node_ty_name.clone())?.get(schema)?;
+                let fields_iter = match node_ty {
+                    ExtendedType::Object(obj) => obj.fields.iter(),
+                    ExtendedType::Interface(int) => int.fields.iter(),
```

Comment:
kind of a nit: It doesn't look like an iterator is the best here, `obj.fields` and `int.fields` are both an IndexMap, we can just use them directly. eg. `fields.get()` instead of `fields_iter.find()` and `fields.keys()` instead of `fields_iter.map()`.

Reply:
I already did this while reviewing to check if it was right 🙈